File: csharp/patterns/src/PrototypePatterns/Creational/Prototype/PrototypePattern.cs
namespace PrototypePatterns.Creational.Prototype;

public sealed class Document(string title, IReadOnlyList<string> tags)
{
    public string Title { get; } = title;
    public List<string> Tags { get; } = [.. tags];

    public Document CloneDocument() => new(Title, Tags);
}

public static class PrototypePattern
{
    public static bool Demo()
    {
        var original = new Document("Pattern Notes", ["draft"]);
        var copy = original.CloneDocument();
        copy.Tags.Add("review");
        return original.Tags.Count == 1 && copy.Tags.Count == 2;
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/creational/prototypepattern/PrototypePattern.java
package org.prototypepatterns.creational.prototypepattern;

import java.util.ArrayList;
import java.util.List;

public final class PrototypePattern {
    private PrototypePattern() {
    }

    public static final class Document {
        public final String title;
        public final List<String> tags;

        public Document(String title, List<String> tags) {
            this.title = title;
            this.tags = tags;
        }

        public Document cloneDocument() {
            return new Document(title, new ArrayList<>(tags));
        }
    }

    public static boolean demo() {
        Document original = new Document("Pattern Notes", new ArrayList<>(List.of("draft")));
        Document copy = original.cloneDocument();
        copy.tags.add("review");
        return original.tags.size() == 1 && copy.tags.size() == 2;
    }
}
